### DISCUSSION TITLE
Add Repeating Form Fields

### DISCUSSION THREAD

**Discussion by msj1542** [Budibase/budibase#6230] [Help]
I am looking for a way to add functionality to a form where the user can click a button and it will add a cloned version of the same field(s), for the addition of additional information, but am not having any success with this.  I am thinking it is maybe because I don't know what the proper terminology for this is, therefore I am not searching in the right place.  Here is a screenshot of the built in Dynamic Filter tool, which happens to have the exact same functionality that I am looking for, in case my description was very unclear. lol.  Thanks! 
![Screenshot 2022-06-07 184914](https://user-images.githubusercontent.com/102364827/172502639-4a49b87c-a7b1-4216-b0a1-0b7515c426c8.jpg)

**Comment by melohagan** [comment]
Hi @msj1542 

I believe this is the same request here: https://github.com/Budibase/budibase/issues/5658

Definitely would be a nice to have!

**Reply by msj1542** [reply]
Thanks!  I think I did run across that link at some point though, which was referring more so to a component/tool that does this, which I agree would be helpful.  I feel like the functionality exists within Budibase though to make this manually, just would obviously require a few more steps than if there were a pre built tool for it, but I just for some reason am having a hard time wrapping my head around the best way of doing this.  To start, not sure if it would it be best through using a single JSON field, or creating a relational table to house the data for these dynamic fields, etc.  Then, I'm also not familiar enough with Budibase yet to know how, or if/ the functionality exists to create a button action that adds/duplicates/clones (or even unhides one), etc., and if so, what the process would be to make it just mirror the layout of them automatically.  I feel like the best way to work this is through a repeater, however, as best I can tell the repeater component is more geared towards pulling/displaying data and not for form functionality, but again, I am still king of green in regards to Budibase's capabilities.  So, any advice or input that maybe even sparks an idea on how this would best work would be greatly appreciated!

**Reply by msj1542** [reply]
**because this EXACT same functionality that I am trying to replicate exists within the platform itself's UI, as shown in the screenshot, but also I have seen it in many other places too, like button action chain settings, or example.  So I figured it may be a pretty simple thing to replicate, but I'm starting to think that there may be way more moving parts to that than what I am considering.

**Reply by melohagan** [reply]
When you put it like that, it does sound like a copy/paste job, but I'm sure there's a fair bit of work to it in reality 😄 

My thinking is that it would be some sort of new component, maybe *Repeatable Field Group*. It is much the same as field group, but has configurable plus/minus buttons for duplicating or removing repeated sections.

**Reply by msj1542** [reply]
I agree, that would be a very convenient component to add, and definitely the easiest way to implement it.  I still feel like there is a way to do it manually though, as least as I very roughly run through the steps in my head and what is needed to make this happen, but my inexperience with Budibase, I think, is what is causing my main hang up here lol.

**Reply by melohagan** [reply]
Well I had a go of sorts here: https://github.com/Budibase/budibase/discussions/5640#discussioncomment-2662603

It uses form steps, but it's limited in that there has to be a max number (which you also need to know ahead of time) of elements that can be added.
You can also only add and remove inputs like a stack data type - last in first out.

You could build it with repeaters as well which would give you greater flexibility (e.g. you could remove inputs at a specific index), but the UX would be fairly gnarly. 
For example it would reload your data every time there was an element added/removed and it would display the "Row Saved/Deleted" toast when adding/removing inputs.

**Reply by msj1542** [reply]
that's interesting.  I feel like with some variation of that I can probably make something work, or at least put something together for the interim that will generate at least some of he results I need.  I appreciate the help!
